src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -62,22 +62,33 @@ async def generate_lockfile(
     python_setup: PythonSetup,
     workspace: Workspace,
 ) -> LockGoal:
-    # TODO: Looking at the transitive closure to generate a single lockfile will not work when we
-    #  have multiple lockfiles supported, via per-tool lockfiles and multiple user lockfiles.
+    # TODO(#12314): Looking at the transitive closure to generate a single lockfile will not work
+    #  when we have multiple lockfiles supported, via per-tool lockfiles and multiple user lockfiles.
     #  Ideally, `./pants lock ::` would mean "regenerate all unique lockfiles", whereas now it
     #  means "generate a single lockfile based on this transitive closure."
     transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest(addresses))
+
+    # TODO(#12314): Likely include "dev dependencies" like MyPy and Pytest, which must not have
+    #  conflicting versions with user requirements. This is a simpler alternative to shading, which
+    #  is likely out of scope for the project. See https://github.com/pantsbuild/pants/issues/9206.
+    #
+    #  We may want to redesign how you set the version and requirements for subsystems in the
+    #  process. Perhaps they should be directly a python_requirement_library, and you use a target
+    #  address? Pytest is a particularly weird case because it's often both a tool you run _and_
+    #  something you import.
+    #
+    #  Make sure to not break https://github.com/pantsbuild/pants/issues/10819.
```

Comment:
While this is a good idea, it seems orthogonal to the current project. We should definitely be able to continue to "concatenate" the inputs artifacts onto the PYTHONPATH before running the tool for now.

Reply:
I don't think it's orthogonal. See https://docs.google.com/document/d/1bCYb0UQZx9a-9tAagydCN_z3826QRvz_3aVnXKSNTJw/edit#heading=h.1nndd74l1gpv for why I think this is broken and about to get much more broken, and discussion of the two proposed fixes.

Reply:
As John commented there:
> We could solve just 1 and leave the other status quo.

... and I agree. The status quo is ok for now, or at least not related to this project.